### PR TITLE
GH-41095: [C++][FS][Azure] Add support for CopyFile with hierarchical namespace support

### PR DESCRIPTION
**Pull request by kou**
### Rationale for this change

We need to add SAS (Shared Access Signatures) token for source URL.

### What changes are included in this PR?

Generate SAS token for source URL.

### Are these changes tested?

Yes.

### Are there any user-facing changes?

Yes.
* GitHub Issue: #41095

### CODE REVIEW

File: cpp/src/arrow/filesystem/azurefs.cc [hunk 1]
```diff
@@ -381,6 +388,24 @@ AzureOptions::MakeDataLakeServiceClient() const {
   return Status::Invalid("AzureOptions doesn't contain a valid auth configuration");
 }
 
+Result<std::string> AzureOptions::GenerateSASToken(
+    Storage::Sas::BlobSasBuilder* builder) const {
+  if (storage_shared_key_credential_) {
+    return builder->GenerateSasToken(*storage_shared_key_credential_);
+  } else {
+    // This part isn't tested. This may not work.
```
Comment (kou) [inline cpp/src/arrow/filesystem/azurefs.cc:396]: @Tom-Newton It seems that we can't use `GetUserDelegationKey()` with account key credential. We may need to other authentication to test `GetUserDelegationKey()`. Do you know how to configure other authentication...?

(I think that we can defer this case to a separated issue.)

File: cpp/src/arrow/filesystem/azurefs.cc [hunk 2]
```diff
@@ -381,6 +388,24 @@ AzureOptions::MakeDataLakeServiceClient() const {
...
+    ARROW_ASSIGN_OR_RAISE(auto client, MakeBlobServiceClient());
```
Comment (felipecrv) [inline cpp/src/arrow/filesystem/azurefs.cc:397]: The caller already has a `BlobServiceClient`, so it's probably a good idea to have a version of `GenerateSASToken` that takes one as parameter.

File: cpp/src/arrow/filesystem/azurefs.cc [hunk 3]
```diff
@@ -2868,8 +2886,19 @@ class AzureFileSystem::Impl {
...
+      Storage::Sas::BlobSasBuilder builder;
+      std::chrono::seconds available_period(60);
+      builder.ExpiresOn = std::chrono::system_clock::now() + available_period;
+      builder.BlobContainerName = src.container;
+      builder.BlobName = src.path;
+      builder.Resource = Storage::Sas::BlobSasResource::Blob;
+      builder.SetPermissions(Storage::Sas::BlobSasPermissions::Read);
+      ARROW_ASSIGN_OR_RAISE(sas_token, options_.GenerateSASToken(&builder));
+    }
+    auto src_url = GetBlobClient(src.container, src.path).GetUrl() + sas_token;
```
Comment (felipecrv) [inline cpp/src/arrow/filesystem/azurefs.cc:2900]: What is a sas token that it can be concatenated just like this to an URL?

Comment (kou) [reply cpp/src/arrow/filesystem/azurefs.cc:2900]: A sample SAS token: `?se=2024-04-18T21:17:19Z&sig=HI/8c...ZExA%3D&sp=r&spr=https&sr=b&sv=2022-11-02`

File: cpp/src/arrow/filesystem/azurefs.cc (hunk 1 above)
Comment (Tom-Newton) [reply cpp/src/arrow/filesystem/azurefs.cc:396]: I'm pretty sure its possible to create a SAS token from an account key but it probably isn't called delegated anymore.

File: cpp/src/arrow/filesystem/azurefs.cc (hunk 2 above)
Comment (kou) [reply cpp/src/arrow/filesystem/azurefs.cc]: OK. I've added it.

File: cpp/src/arrow/filesystem/azurefs.cc (hunk 1 above)
Comment (kou) [reply cpp/src/arrow/filesystem/azurefs.cc:396]: Yes. We can create a SAS token from an account key. The above `builder->GenerateSasToken(*storage_shared_key_credential_)` does it.

Comment (Tom-Newton) [reply cpp/src/arrow/filesystem/azurefs.cc:396]: Sorry I misunderstood the intention (I haven't found time to review this). 

Configuring other auths in testing might be a bit difficult. For Azurite I think the first challenge is configuring `https`, which I have struggled with before. Beyond that I think there are 2 viable options offered by Azure:
1. Service principal
    1. Username and password for a service account
    2. Surprisingly complicated to create
1. Azure CLI 
    1. Authenticates as your user
    2. Requires installing the Azure CLI and running `az login`, then completing the login in your browser.

File: cpp/src/arrow/filesystem/azurefs.cc (hunk 3 above)
Comment (felipecrv) [reply cpp/src/arrow/filesystem/azurefs.cc:2900]: Ok. It comes with the `?` already.

**Review by felipecrv — APPROVED** [review]

File: cpp/src/arrow/filesystem/azurefs.cc (hunk 1 above)
Comment (kou) [reply cpp/src/arrow/filesystem/azurefs.cc:396]: Thanks.
I tried `az login` and `ConfigureDefaultCredential()` but some operations such as `CreateDir()` was blocked...
It seems that we need to implement GH-39344 to test this part.

I'll open a new issue for this part and defer this part in this PR.

Comment (kou) [reply cpp/src/arrow/filesystem/azurefs.cc]: #41315

Comment (Tom-Newton) [reply cpp/src/arrow/filesystem/azurefs.cc]: > I tried az login and ConfigureDefaultCredential() but some operations such as CreateDir() was blocked...
It seems that we need to implement https://github.com/apache/arrow/issues/39344 to test this part.

This seems a bit suspicious. Default credential does include CLI auth...

Comment (kou) [reply cpp/src/arrow/filesystem/azurefs.cc]: Let's look at this in #41315.

**Comment by kou** [comment]
I'll merge this.

**Pull request merged**